opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -2652,29 +2652,54 @@ namespace Opm {
+      function must be called for all processors - even if they do not have the
+      well in question.
+    */
     template<typename TypeTag>
     void
     BlackoilWellModel<TypeTag>::
-    updateEclWell(const int timeStepIdx, const int well_index)
+    updateEclWell(int timeStepIdx, const std::optional<int>& well_index)
     {
         const auto& schedule = this->ebosSimulator_.vanguard().schedule();
-        const auto& wname = this->wells_ecl_[well_index].name();
-        this->wells_ecl_[well_index] = schedule.getWell(wname, timeStepIdx);
+        if (well_index.has_value()) {
+            auto wi = well_index.value();
+            const auto& wname = this->wells_ecl_[wi].name();
+            this->wells_ecl_[wi] = schedule.getWell(wname, timeStepIdx);
+
+            const auto& well = this->wells_ecl_[wi];
+            auto& pd     = this->well_perf_data_[wi];
+            auto  pdIter = pd.begin();
+            for (const auto& conn : well.getConnections()) {
+                if (conn.state() != Connection::State::SHUT) {
+                    pdIter->connection_transmissibility_factor = conn.CF();
+                    ++pdIter;
+                }
+            }
 
-        const auto& well = this->wells_ecl_[well_index];
-        auto& pd     = this->well_perf_data_[well_index];
-        auto  pdIter = pd.begin();
-        for (const auto& conn : well.getConnections()) {
-            if (conn.state() != Connection::State::SHUT) {
-                pdIter->connection_transmissibility_factor = conn.CF();
-                ++pdIter;
+            this->well_state_.updateStatus(wi, well.getStatus());
+            if (well.getStatus() == Well::Status::OPEN) {
+                printf("Opening well: injector: %d  %s\n", well.isInjector(), well.name().c_str());
+                if (well.isInjector()) {
+                    auto controls = this->well_state_.currentInjectionControls();
+                    printf("Control: %d -> ", static_cast<int>(controls[wi]));
+                    if (controls[wi] == Well::InjectorCMode::CMODE_UNDEFINED)
+                        controls[wi] = well.injection_cmode();
+                    printf("%d \n", static_cast<int>(controls[wi]));
+                } else {
+                    auto controls = this->well_state_.currentProductionControls();
+                    if (controls[wi] == Well::ProducerCMode::CMODE_UNDEFINED)
+                        controls[wi] = well.production_cmode();
+                }
             }
-        }
 
-        this->well_state_.resetConnectionTransFactors(well_index, pd);
-        this->prod_index_calc_[well_index].reInit(well);
+            this->well_state_.resetConnectionTransFactors(wi, pd);
+            this->prod_index_calc_[wi].reInit(well);
+        }
+        const auto& comm = this->ebosSimulator_.vanguard().grid().comm();
+        this->well_state_.updateGlobalIsGrup(schedule, timeStepIdx, comm);
     }
```

Comment:
I suggest to do a full `updateAndCommunicateGroupData` here not just update the global Grup.

Reply:
Sure you need the `updateGlobalIsGrup`. I removed it get the same results. 
The `applyActions` is called in `endTimeStep` and `updateAndCommunicateGroupData` (where `updateGlobalIsGrup` is called) is called in `beginTimeStep`. Nothing well related is happening in between.

Reply:
No I am not sure; I was trying so many things. Your argument is good - I will remove it.